contracts/pcv/ERC4626/ERC4626PCVDeposit.sol

```diff
@@ -0,0 +1,272 @@
+        uint256 amount
+    ) external onlyPCVController globalLock(2) {
+        int256 startingRecordedBalance = lastRecordedBalance.toInt256();
+
+        /// compute profit from interest accrued and emit an event
+        _recordPNL();
+
+        _withdraw(to, amount);
+
+        int256 endingRecordedBalance = lastRecordedBalance.toInt256();
+
+        _updateOracle(endingRecordedBalance - startingRecordedBalance);
+    }
+
+    /// @notice withdraw all withdrawable tokens from the vault
+    /// This is slightly different than a withdrawAll function
+    /// as the vault might not let the PCVDeposit to redeem all
+    /// its shares directly. Example if a Maple Vault does not have
+    /// enough liquidity because most of it has been borrowed
+    /// @param to the address PCV will be sent to
+    function withdrawMax(address to) external onlyPCVController globalLock(2) {
+        int256 startingRecordedBalance = lastRecordedBalance.toInt256();
+
+        /// compute profit from interest accrued and emit an event
+        _recordPNL();
+
+        /// withdraw last recorded amount as this was updated in record pnl
+        _withdraw(to, withdrawableBalance());
+
+        int256 endingRecordedBalance = lastRecordedBalance.toInt256();
+
+        _updateOracle(endingRecordedBalance - startingRecordedBalance);
+    }
+
+    /// ------------------------------------------
+    /// ------------- Helper Methods -------------
+    /// ------------------------------------------
+
+    /// @notice update the PCVOracle if delta is not 0
+    function _updateOracle(int256 delta) private {
+        if (delta != 0) {
+            isLiquid
+                ? _liquidPcvOracleHook(delta)
+                : _illiquidPcvOracleHook(delta);
+        }
+    }
+
+    /// @notice helper function to avoid repeated code in withdraw and withdrawMax
+    /// anytime this function is called it is by an external function in this smart contract
+    /// with a reentrancy guard. This ensures lastRecordedBalance never desynchronizes.
+    /// @param to recipient of withdraw funds
+    /// @param amount to withdraw
+    function _withdraw(address to, uint256 amount) private {
+        /// ------ Effects ------
+
+        /// update last recorded balance amount
+        /// if more than is owned is withdrawn, this line will revert
+        /// this line of code is both a check, and an effect
+        lastRecordedBalance -= amount;
+
+        /// ------ Interactions ------
+
+        IERC4626(vault).withdraw(
+            amount,
+            to, // receiver
+            address(this) // owner of the tokens
+        );
+
+        emit Withdrawal(msg.sender, to, amount);
+    }
+
+    /// @notice records how much profit or loss has been accrued
+    /// since the last call and emits an event with all profit or loss received.
+    /// Updates the lastRecordedBalance to include all realized profits or losses.
+    function _recordPNL() private {
+        /// ------ Check ------
+
```

Comment:
missing a hook here that can be overridden in child classes to poke the external venue and accrue interest if necessary. hook should start off empty, internal and virtual so that children can override and define new behavior. this would allow integration with morpho or compound directly using this vault so that `accrue` or `updateP2PIndexes` can be called.

Reply:
ok, maybe this should be a discussion as we might not need this.